phe/encoding.py

```diff
@@ -186,7 +188,27 @@ def encode(cls, public_key, scalar, precision=None, max_exponent=None):
         else:
             exponent = min(max_exponent, prec_exponent)
 
-        int_rep = int(round(scalar * pow(cls.BASE, -exponent)))
+        # Below is just `int_rep = round(scalar * pow(BASE, -exponent))`
+        # but accouting for all the cases.
+        if isinstance(scalar, int):
+            if exponent <= 0:
+                # Just integer multiplication. This is exact.
+                int_rep = scalar * cls.BASE ** -exponent
+            else:
+                # Integer division with correct rounding.
+                # Can't just to `scalar * cls.BASE ** -exponent` as it
+                # would coerce scalar to float, potentially overflowing.
+                int_rep = int_div_round(scalar, cls.BASE ** exponent)
+        elif isinstance(scalar, float):
+            # It is possible for scalar and BASE ** -exponent to not be
+            # representable as float, even when the encoded value is
+            # a perfectly valid float. We represent scalar as integers
+            # and avoid coercing to float.
```

Comment:
Possible to misunderstand your comment here as the `scalar` you mention is different from the current scoped variable `scalar` which at this point **is** a float.